google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/BitReversedSequenceStyleGenerator.java

```diff
@@ -154,14 +157,45 @@ protected DatabaseStructure buildTableStructure(
         type.getReturnedClass());
   }
 
+  private static final int MAX_ATTEMPTS = 100;
+
   /**
    * Generates a new ID. This uses the normal sequence strategy, but the returned ID is bit-reversed
    * before it is returned to the application.
    */
   @Override
   public Serializable generate(SharedSessionContractImplementor session, Object object)
       throws HibernateException {
-    Serializable id = generateBaseValue(session, object);
+    Serializable id;
+    int attempts = 0;
+    // Loop to retry the transaction that updates the table-backed sequence if it fails due to a
+    // concurrent modification error. This can happen if multiple entities are using the same
+    // sequence for generated primary keys.
+    while (true) {
+      try {
+        id = generateBaseValue(session, object);
+        break;
+      } catch (GenericJDBCException exception) {
+        JdbcAbortedDueToConcurrentModificationException aborted;
+        if (exception.getSQLException()
+            instanceof JdbcAbortedDueToConcurrentModificationException) {
+          aborted = (JdbcAbortedDueToConcurrentModificationException) exception.getSQLException();
+        } else {
+          throw exception;
+        }
+        attempts++;
+        if (attempts == MAX_ATTEMPTS) {
+          throw exception;
+        }
+        try {
+          Thread.sleep(aborted.getCause().getRetryDelayInMillis());
+        } catch (InterruptedException interruptedException) {
+          Thread.currentThread().interrupt();
+          throw new IdentifierGenerationException("Interrupted while trying to generate a new ID",
+              interruptedException);
+        }
+      }
+    }
     if (id instanceof Long) {
```

Comment:
Could you add a test case to generate a sequence that is not an instance of `Long`?

Reply:
Added a test for a generator that returns an Integer and one that returns a String.